app-api/src/main/java/org/openmrs/eip/app/CustomDatabaseOffsetBackingStore.java

```diff
@@ -0,0 +1,135 @@
+import org.openmrs.eip.app.management.entity.DebeziumOffset;
+import org.openmrs.eip.app.management.repository.DebeziumOffsetRepository;
+import org.openmrs.eip.app.util.DebeziumOffsetUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import io.debezium.config.Instantiator;
+import io.debezium.embedded.EmbeddedEngine;
+
+@Component
+public class CustomDatabaseOffsetBackingStore extends MemoryOffsetBackingStore {
+	
+	private static final Logger log = LoggerFactory.getLogger(CustomDatabaseOffsetBackingStore.class);
+	
+	private static DebeziumOffsetRepository debeziumOffsetRepository;
+	
+	private String engineName;
+	
+	private Converter valueConverter;
+	
+	public CustomDatabaseOffsetBackingStore() {
+	}
+	
+	@Autowired
+	public CustomDatabaseOffsetBackingStore(DebeziumOffsetRepository debeziumOffsetRepository) {
+		CustomDatabaseOffsetBackingStore.debeziumOffsetRepository = debeziumOffsetRepository;
+	}
+	
+	@Override
+	public synchronized void start() {
+		super.start();
+		load();
+	}
+	
+	/**
+	 * @see {@link io.debezium.embedded.EmbeddedEngine#EmbeddedEngine()}
+	 */
+	@Override
+	public void configure(WorkerConfig config) {
+		Map<String, Object> originals = config.originals();
+		
+		this.engineName = (String) originals.get(EmbeddedEngine.ENGINE_NAME.name());
+		String valueConverterClassName = (String) originals.get(WorkerConfig.VALUE_CONVERTER_CLASS_CONFIG);
+		
+		Map<String, Object> valueConverterOriginalConfs = config
+		        .originalsWithPrefix(WorkerConfig.INTERNAL_VALUE_CONVERTER_CLASS_CONFIG + ".", true);
+		valueConverterOriginalConfs.put("schemas.enable", false);
+		
+		this.valueConverter = Instantiator.getInstance(valueConverterClassName, () -> this.getClass().getClassLoader(),
+		    null);
+		this.valueConverter.configure(valueConverterOriginalConfs, false);
+	}
+	
+	/**
+	 * @see
+	 */
+	private void load() {
+		long offsets = debeziumOffsetRepository.countByEnabledTrue();
+		
+		if (offsets > 0) {
+			log.info("Loading offset from the database");
+			DebeziumOffset offset = debeziumOffsetRepository.findFirstByEnabledTrueOrderByDateCreatedDesc();
+			
+			if (log.isDebugEnabled()) {
+				log.debug("Loaded offset: {}", offset);
+			}
+			
+			Map<byte[], byte[]> raw = (Map<byte[], byte[]>) SerializationUtils.deserialize(offset.getData());
+			data = DebeziumOffsetUtil.offsetRawToData(raw);
+			
+		} else {
+			log.info("No offset was found in the database");
+		}
+	}
+	
+	@Override
+	public void save() {
+		CustomOffsetBackingStore.save(() -> {
+			if (log.isDebugEnabled()) {
+				log.debug("Saving offset to the database");
+			}
+			
+			DebeziumOffset offset = new DebeziumOffset();
+			offset.setDateCreated(new Date());
+			offset.setData(SerializationUtils.serialize((Serializable) DebeziumOffsetUtil.offsetDataToRaw(data)));
```

Comment:
We also though that in the first instance, but we could end adding extra complexity during the conversion from/to the database. Additionally, in case of some change in offset structure (new column, etc), we will need to update our conversion logic to support that.
But, we can implement in that way, because we already implemented the logic to capture a single map value from the offset data.